### PR TITLE
API changes to allow presigned URLs to download private datasets when logged in with Auth0

### PR DESCRIPTION
**Pull request by kennyatpillar**
This is a DRAFT PR because discovery_api is moving to the umbrella project smartcitiesdata.

We'd still like feedback here and will address it as we move our changes into smartcitiesdata/apps/discovery_api after that work is complete.

### CODE REVIEW

File: lib/discovery_api_web/controllers/data_controller.ex [hunk 1]
```diff
@@ -38,44 +39,11 @@ defmodule DiscoveryApiWeb.DataController do
...
-        dataset_name: dataset_name,
-        schema: schema
-      })
-
-    resp_as_stream(conn, rendered_data_stream, format, dataset_id)
+  def download_presigned_url(conn, params) do
+    expires_in_seconds = Application.get_env(:discovery_api, :download_link_expire_seconds)
+    expires = DateTime.utc_now() |> DateTime.add(expires_in_seconds, :second) |> DateTime.to_unix()
+    hmac_token = HmacToken.create_hmac_token(params["dataset_id"], expires)
+    json(conn, "/dataset/#{params["dataset_id"]}/download?key=#{hmac_token}&expires=#{expires}")
```
Comment (ManApart) [inline lib/discovery_api_web/controllers/data_controller.ex]: This just returns a straight string, which works for us on the front end. We _could_ make this return an actual json object, something like below, but we're not sure the JSR is there. We'd be curious to hear thoughts.

```
{
"downloadUrl": "urlurlurlurlurl"
}
```

File: config/config.exs [hunk 2]
```diff
@@ -19,7 +19,9 @@ config :logger, :console,
 config :discovery_api,
   auth_provider: "default",
   collector: StreamingMetrics.PrometheusMetricCollector,
-  hsts_enabled: true
+  presign_key: "test_presign_key",
```
Comment (jmorris-pillar) [inline config/config.exs]: override this in runtime.exs

File: lib/discovery_api_web/controllers/data_controller.ex (hunk 1 above)
Comment (jmorris-pillar) [reply lib/discovery_api_web/controllers/data_controller.ex]: I dont care as long as the UI folks dont care about using diff content type responses on async calls

File: lib/discovery_api_web/controllers/data_download_controller.ex [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+  def fetch_file(
+        %{assigns: %{model: %{private: true, id: dataset_id}}} = conn,
+        %{"expires" => expires, "key" => key},
+        format
+      ) do
+    dataset_name = conn.assigns.model.systemName
+    schema = conn.assigns.model.schema
+
+    parsed_expires = ParamUtils.safely_parse_int(expires)
```
Comment (jmorris-pillar) [inline lib/discovery_api_web/controllers/data_download_controller.ex]: why do we parse this diff here

Comment (jmorris-pillar) [reply lib/discovery_api_web/controllers/data_download_controller.ex]: Than in the above method that is

File: lib/discovery_api_web/router.ex [hunk 4]
```diff
@@ -72,12 +72,17 @@ defmodule DiscoveryApiWeb.Router do
...
     get("/dataset/:dataset_id/query", DataController, :query)
-    get("/organization/:org_name/dataset/:dataset_name/download", DataController, :fetch_file)
-    get("/dataset/:dataset_id/download", DataController, :fetch_file)
+    get("/dataset/:dataset_id/presign_url", DataController, :download_presigned_url)
 
     resources("/visualization", VisualizationController, only: [:show])
   end
 
+  scope "/api/v1", DiscoveryApiWeb do
+    pipe_through([:reject_cookies_from_ajax, :add_user_details, :global_headers])
```
Comment (LtChae) [inline lib/discovery_api_web/router.ex]: Could HMAC handling be taken care of with a plug here?

File: config/integration.exs [hunk 5]
```diff
@@ -3,7 +3,7 @@ aws_region = "us-west-2"
 host = "localhost"
 endpoints = [{to_charlist(host), 9092}]
 
-config :discovery_api, DiscoveryApiWeb.Endpoint, url: [host: "data.integrationtests.example.com", port: 80]
+config :discovery_api, DiscoveryApiWeb.Endpoint, url: [scheme: "https", host: "data.integrationtests.example.com", port: 80]
```
Comment (ManApart) [inline config/integration.exs]: D?id you guys add this in runtime as well?

File: lib/discovery_api_web/router.ex [hunk 6]
```diff
@@ -72,12 +72,17 @@ defmodule DiscoveryApiWeb.Router do
     get("/dataset/:dataset_id/preview", DataController, :fetch_preview)
     get("/organization/:org_name/dataset/:dataset_name/query", DataController, :query)
     get("/dataset/:dataset_id/query", DataController, :query)
-    get("/organization/:org_name/dataset/:dataset_name/download", DataController, :fetch_file)
-    get("/dataset/:dataset_id/download", DataController, :fetch_file)
+    get("/dataset/:dataset_id/presign_url", DataController, :download_presigned_url)
```
[the remark this answers is not in the document]
Comment (kennyatpillar) [reply lib/discovery_api_web/router.ex]: Agreed, made this change

**Pull request merged**